templates/template.rb

```diff
@@ -0,0 +1,119 @@
+      attr_reader :parent_c_constant_name
+      attr_reader :fields
+
+      def initialize(yaml)
+        @full_name = yaml["name"]
+        @name = yaml["name"].split("::").last
+        @c_function_name = if @full_name =~ /^RBS::Types/
+          name = @full_name.gsub("::", "_").underscore
+          name.gsub("_types_", "_") # TODO: should we rename these methods?
+        else
+          name = @full_name.gsub("::", "_").underscore
+          name.gsub("_declarations_", "_decl_") # TODO: should we rename these methods?
+        end
+        @c_constant_name = @full_name.gsub("::", "_")
+        @fields = yaml["fields"].map { |field| Field.new(field) }
+        @parent_c_constant_name = @full_name.split("::")[0..-2].join("::").gsub("::", "_")
+      end
+    end
+
+    class << self
+      def render(out_file)
+        filepath = "templates/#{out_file}.erb"
+        template = File.expand_path("../#{filepath}", __dir__)
+
+        erb = read_template(template)
+        extension = File.extname(filepath.gsub(".erb", ""))
+
+        heading = <<~HEADING
+          /*----------------------------------------------------------------------------*/
+          /* This file is generated by the templates/template.rb script and should not  */
+          /* be modified manually.                                                      */
+          /* To change the template see                                                 */
+          /* #{filepath + " " * (74 - filepath.size) } */
+          /*----------------------------------------------------------------------------*/
+        HEADING
+
+        write_to = File.expand_path("../#{out_file}", __dir__)
+        contents = heading + "\n" + erb.result_with_hash(locals)
+
+        if (extension == ".c" || extension == ".h") && !contents.ascii_only?
+          # Enforce that we only have ASCII characters here. This is necessary
+          # for non-UTF-8 locales that only allow ASCII characters in C source
+          # files.
+          contents.each_line.with_index(1) do |line, line_number|
+            raise "Non-ASCII character on line #{line_number} of #{write_to}" unless line.ascii_only?
+          end
+        end
+
+        FileUtils.mkdir_p(File.dirname(write_to))
+        File.write(write_to, contents)
+      end
+
+      private
+
+      def read_template(filepath)
+        template = File.read(filepath, encoding: Encoding::UTF_8)
+        erb = erb(template)
+        erb.filename = filepath
+        erb
+      end
+
+      def erb(template)
+        ERB.new(template, trim_mode: "-")
+      end
+
+      def locals
+        config = YAML.load_file(File.expand_path("../config.yml", __dir__))
+        {
+          nodes: config.fetch("nodes").map { |node| Type.new(node) }.sort_by(&:full_name),
+        }
+      end
+    end
+  end
+end
+
+class String
+  # From https://github.com/rails/rails/blob/a1f6a13f691e0929d40b7e1b1e0d31aa69778128/activesupport/lib/active_support/inflector/methods.rb#L99
```

Comment:
I just realized that `activesupport` is in the Gemfile.lock, should we use it instead?